### ISSUE TITLE
Improve --fullpath and --match-d root path

### ISSUE THREAD

**Issue by PMExtra** [AlDanial/cloc#731]
For now, `--fullpath` or `--match-d` options will make cloc evaluate the regex with the physical full path. Which is not interesting to users.

If I want to match `./foo/src/` but neither `./src/` nor `./next/foo/src/`,
I have to set `--match-d="^$(pwd)/foo/src/"`.

Also, when I want to work with git diff, it is an unqualified disaster. Because the diff will make temporary directories that hard to get the name.

I think we can make the cloc work path as a virtual root to solve these problems, just like git.

It makes we can use `match-d=^/foo/src/` to match only the certain path relative to the project.

**Comment by martinvonwittich** [comment]
I also just stumbled over this and was totally confused how to anchor the regex - I wanted to exclude `./path/to/foo/`, but had no success with  `--not-match-d` `^\./path/to/foo`, `^/path/to/foo` and `^path/to/foo`. I was also confused by `--fullpath`; the manpage says:

```
       --not-match-d=REGEX
           Count all files except in directories matching the Perl regex.
           Only the trailing directory name is compared, for example, when
           counting in "/usr/local/lib", only "lib" is compared to the regex.
           Add --fullpath to compare parent directories to the regex.  Do not
           include file path separators at the beginning or end of the regex.
```

But `--fullpath` claims only to influence `match-f`, and doesn't say anything about `match-d`:

```
       --fullpath
           Modifies the behavior of --match-f or --not-match-f to include the
           file's path in the regex, not just the file's basename.  (This does
           not expand each file to include its absolute path, instead it uses
           as much of the path as is passed in to cloc.)
```

And funnily enough, `--match-d='path/to/foo` *without* `--fullpath` excludes everything outside of `/path/to/foo`, so apparently it does work without `--fullpath`?

To figure out how to anchor my regex, I had to put some debug code into `cloc` to see what's actually being matched here:

```
    if ($opt_match_d    ) { use DDP; p $Dir; return unless $Dir =~ m{$opt_match_d}     }
```

and I was dismayed to learn that it matches the absolute path, including the path to my home directory.

**Comment by AlDanial** [comment]
I agree this needs work (both the code and the documentation).  Basing `--fullpath` off the directory from which cloc was invoked sounds reasonable.  Many switches will be affected so a fix will be a while in coming.

**Comment by AlDanial** [comment]
@PMExtra The intent of `--fullpath` was always meant to be relative to the directory from which cloc was run.  In that sense it is poorly named; `--filepath` would have been a better choice.  I still haven't finished testing the `--match-d` branches, so the current commit may not interest you.

@martinvonwittich There was a problem with how `--not-match-d` was implemented.  Your combination `--not-match-d ^\./path/to/foo` should have given the behavior you desired .  Please try 18062b5 to see if `--not-match-d` works as you expect.

**Comment by AlDanial** [comment]
@PMExtra try 39f3b9e.  With that fix  `--match-d="^\./foo/src/"` should give you the behavior you want (without `--fullpath`).
